searx/version.py

```diff
@@ -1,26 +1,178 @@
+    try:
+        ref = call_subprocess_safe("git rev-parse --abbrev-ref @{upstream}")
+    except ShellException:
+        ref = call_subprocess_safe("git rev-parse --abbrev-ref master@{upstream}")
+    origin, git_branch = ref.split("/", 1)
+    git_url = call_subprocess_safe(f"git remote get-url {origin}")
+    return get_public_git_url(git_url), git_branch
+
+
+def get_tag_version(tag_version):
+    if re.match(r"v[0-9]+\.[0-9]+\.[0-9]+", tag_version):
+        # tag version "v1.0.0" becomes "1.0.0" (without the v)
+        # other patterns are kept untouched
+        return tag_version[1:]
+    return tag_version
+
+
+def get_git_version():
+    try:
+        tag = call_subprocess_safe("git describe HEAD")
+        tag_split = tag.split("-")
+        if len(tag_split) == 3:
+            # match "<tag>-<distance>-g<commit>"
+            # example: "v1.0.0-590-g0686e274"
+            tag_version, tag_distance, tag_commit = tag_split
+            tag_version = get_tag_version(tag_version)  # remove prefix "v"
+            tag_commit = tag_commit[1:]  # remove prefix "g"
+            # set git_version to "1.0.0-590-0686e274"
+            git_version = f"{tag_version}-{tag_distance}-{tag_commit}"
+        else:
+            # HEAD is on tag name, example: tag = "v1.0.1"
+            git_version = get_tag_version(tag)
+            tag_version = git_version
+    except ShellException:
+        # fall back to "YYYY.MM.DD.Hash" if there is no tag at all
+        git_version = call_subprocess_safe(r"git show -s --format='%as-%h'")
+        # PEP 440: replace - with .
+        git_version = git_version.replace("-", ".")
+        tag_version = git_version
+
+    # skip 'git update-index -q --refresh' to avoid any git repository change
+
+    # add "-dirty" suffix if there are uncommited changes
+    diff = call_subprocess_safe("git diff-index --name-only HEAD --")
+    if diff:
+        git_version += "-dirty"
+    return git_version, tag_version
+
+
+try:
+    from searx.version_frozen import VERSION_STRING, VERSION_TAG, GIT_URL, GIT_BRANCH
+except ImportError:
+    try:
+        VERSION_STRING, VERSION_TAG = get_git_version()
+    except ShellException as e:
+        logger.error("Error while getting the version: %s", e.stderr)
+        VERSION_STRING = "1.0.0"
+        VERSION_TAG = "1.0.0"
+    try:
+        GIT_URL, GIT_BRANCH = get_git_url_and_branch()
+    except ShellException as e:
+        logger.error("Error while getting the git URL & branch: %s", e.stderr)
+        GIT_URL = "unknow"
+        GIT_BRANCH = "unknow"
+
+
+logger.info("version: %s", VERSION_STRING)
+
+if __name__ == "__main__":
+    import sys
+
+    if len(sys.argv) >= 2 and sys.argv[1] == "freeze":
+        # freeze the version (to create an archive outside a git repository)
+        # Looks at the current index and checks to see if merges or updates are needed by checking stat() information.
+        call_subprocess_safe("git update-index -q --refresh")
```

Comment:
For my understanding, for what is this command needed?

Reply:
See commit 2d60d80fe7cf04d50933c6e9469b1f625111309d

This is copy / paste from the existing code (from fbe40001d35ea5cf74c92f35d76c642a480a5c9f ):
https://github.com/searxng/searxng/blob/feef577f996b00c4b4b610b6e3a17e2183d20927/manage#L235-L240

There is no reason to use the low level and "git diff --quiet" makes less syscalls (so most probably fast).

Reply:
Updated